src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+            {
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocations) in projectCallSites)
+                {
+                    modifiedSolution = await ModifyDocumentInvocationsAndIntroduceParameterAsync(compilation,
+                        originalDocument, document, modifiedSolution, methodSymbol, containingMethod, expressionCopy,
+                        expression, allOccurrences, parameterName, invocations, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter.
+        /// </summary>
+        public static async Task<Dictionary<TIdentifierNameSyntax, IParameterSymbol>> MapExpressionToParametersAsync(
+            Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IParameterSymbol parameterSymbol)
+                {
+                    nameToParameterDict.Add(variable, parameterSymbol);
+                }
+            }
+
+            return nameToParameterDict;
+        }
+
+        /// <summary>
+        /// Determines if the expression is something that should have code actions displayed for it.
+        /// Depends upon the identifiers in the expression mapping back to parameters.
+        /// Does not handle params parameters.
+        /// </summary>
+        private static async Task<(bool isParameterized, bool hasOptionalParameter)> ShouldExpressionDisplayCodeActionAsync(
+            Document document, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            CancellationToken cancellationToken)
+        {
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var hasOptionalParameter = false;
+
+            foreach (var parameter in methodSymbol.Parameters)
+            {
+                if (parameter.HasExplicitDefaultValue)
+                {
+                    hasOptionalParameter = true;
+                }
+            }
+
+            foreach (var variable in variablesInExpression)
```

Comment:
you are walking all identifiers.  that means if you have `p.X` it will hit `p` and `X`.  `p` may be a parameter, but `X` could be something else.  it looks like you bail out there, but i can't tell why.

Reply:
Is there a better way to determine how to filter out expressions with local variables? Basically the purpose was that if any identifier did not map back to a parameter, then dont show any code actions

Reply:
yes.  check if you bind to any ILocalSymbol or IRangeVariable symbols.

Reply:
alterantively, you need to make sure you're looking at stand-alone identifiers.  i.e. if you have `a.b` you look only at `a`, not `b`.  we need tests here because currently it seems like you won't work on basic complex expressions like `a.b`.

Reply:
I think I cover that case now? I don't return false anymore if the identifiers dont map back to a parameter, but in the MapExpressionToParameters method I iterate much the same way to create a dictionary of identifier to parameter. Since, in a.b, b doesn't map to a parameter, then it will be ignored in the mapping and ignored when creating the new expression. Is there another way to completely isolate the stand-alone identifiers?